File: server/src/services/pairingService.ts
import { TableClient, TableEntity } from '@azure/data-tables';

export interface Pair {
  person1: string;
  person2: string;
  department1: string;
  department2: string;
}

export interface Participant {
  name: string;
  department: string;
}

export class PairingService {
  private tableClient: TableClient | null;
  private pairingHistory: Map<string, Set<string>> = new Map();
  private mockStorage: Map<string, TableEntity[]> = new Map();

  constructor(tableClient: TableClient | null) {
    this.tableClient = tableClient;
  }

  async generatePairings(participants: Participant[], runDate: string): Promise<Pair[]> {
    // Shuffle participants
    const shuffled = [...participants].sort(() => Math.random() - 0.5);
    const pairs: Pair[] = [];

    // Load existing pairs to avoid duplicates
    await this.loadPairingHistory();

    // Generate pairs ensuring no repeats
    for (let i = 0; i < shuffled.length - 1; i += 2) {
      const p1 = shuffled[i];
      const p2 = shuffled[i + 1];

      const pairKey = this.getPairKey(p1.name, p2.name);

      // Check if this pair already exists
      if (!this.pairingHistory.has(pairKey)) {
        pairs.push({
          person1: p1.name,
          person2: p2.name,
          department1: p1.department,
          department2: p2.department,
        });

        this.pairingHistory.set(pairKey, new Set([p1.name, p2.name]));
      }
    }

    // Store in Azure Table Storage
    await this.storePairings(pairs, runDate);

    return pairs;
  }

  async getPairings(runDate: string): Promise<Pair[]> {
    const pairs: Pair[] = [];
    try {
      if (this.tableClient) {
        const listEntitiesIterator = this.tableClient.listEntities<TableEntity>({
          queryOptions: { filter: `runDate eq '${runDate}'` }
        });

        for await (const entity of listEntitiesIterator) {
          pairs.push({
            person1: entity.person1 as string,
            person2: entity.person2 as string,
            department1: entity.department1 as string,
            department2: entity.department2 as string,
          });
        }
      } else {
        // Use mock storage
        const entities = this.mockStorage.get(runDate) || [];
        entities.forEach(entity => {
          pairs.push({
            person1: entity.person1 as string,
            person2: entity.person2 as string,
            department1: entity.department1 as string,
            department2: entity.department2 as string,
          });
        });
      }
    } catch (error) {
      console.error('Error retrieving pairings:', error);
    }
    return pairs;
  }

  async resetPairings(): Promise<void> {
    try {
      if (this.tableClient) {
        const listEntitiesIterator = this.tableClient.listEntities<TableEntity>();
        for await (const entity of listEntitiesIterator) {
          await this.tableClient.deleteEntity(entity.partitionKey as string, entity.rowKey as string);
        }
      }
      this.mockStorage.clear();
      this.pairingHistory.clear();
    } catch (error) {
      console.error('Error resetting pairings:', error);
      throw error;
    }
  }

  async getExhaustionPercentageAsync(): Promise<number> {
    try {
      let count = 0;
      if (this.tableClient) {
        const listEntitiesIterator = this.tableClient.listEntities<TableEntity>();
        for await (const entity of listEntitiesIterator) {
          count++;
        }
      } else {
        // Use mock storage
        for (const entities of this.mockStorage.values()) {
          count += entities.length;
        }
      }
      // Simple calculation: count / (estimated max possible pairs)
      return Math.min((count / 100) * 100, 100);
    } catch (error) {
      console.error('Error calculating exhaustion:', error);
      return 0;
    }
  }

  getExhaustionPercentage(): number {
    // Sync version for quick calculation
    const maxPossiblePairs = 100; // Estimated
    const usedPairs = this.pairingHistory.size;
    return Math.min((usedPairs / maxPossiblePairs) * 100, 100);
  }

  async exportPairingsCsv(runDate: string): Promise<string> {
    const pairs = await this.getPairings(runDate);
    let csv = 'Person 1,Department 1,Person 2,Department 2\n';
    for (const pair of pairs) {
      csv += `"${pair.person1}","${pair.department1}","${pair.person2}","${pair.department2}"\n`;
    }
    return csv;
  }

  private getPairKey(name1: string, name2: string): string {
    const names = [name1, name2].sort();
    return `${names[0]}|${names[1]}`;
  }

  private async storePairings(pairs: Pair[], runDate: string): Promise<void> {
    for (let i = 0; i < pairs.length; i++) {
      const pair = pairs[i];
      const entity: TableEntity = {
        partitionKey: runDate,
        rowKey: `pair-${i}`,
        person1: pair.person1,
        person2: pair.person2,
        department1: pair.department1,
        department2: pair.department2,
        runDate: runDate,
        timestamp: new Date(),
      };
      
      if (this.tableClient) {
        try {
          await this.tableClient.createEntity(entity);
        } catch (error) {
          console.error('Error storing pair:', error);
          // Fall back to mock storage
          this.storeMockEntity(entity);
        }
      } else {
        this.storeMockEntity(entity);
      }
    }
  }

  private storeMockEntity(entity: TableEntity): void {
    const key = `${entity.partitionKey}|${entity.rowKey}`;
    if (!this.mockStorage.has(entity.partitionKey as string)) {
      this.mockStorage.set(entity.partitionKey as string, []);
    }
    this.mockStorage.get(entity.partitionKey as string)!.push(entity);
  }

  private async loadPairingHistory(): Promise<void> {
    try {
      if (this.tableClient) {
        const listEntitiesIterator = this.tableClient.listEntities<TableEntity>();
        for await (const entity of listEntitiesIterator) {
          const pairKey = this.getPairKey(entity.person1 as string, entity.person2 as string);
          this.pairingHistory.set(pairKey, new Set([entity.person1 as string, entity.person2 as string]));
        }
      } else {
        // Use mock storage
        for (const entities of this.mockStorage.values()) {
          entities.forEach(entity => {
            const pairKey = this.getPairKey(entity.person1 as string, entity.person2 as string);
            this.pairingHistory.set(pairKey, new Set([entity.person1 as string, entity.person2 as string]));
          });
        }
      }
    } catch (error) {
      console.log('No existing pairings found or table does not exist yet');
    }
  }
}

File: server/Services/PairingService.cs
using Azure.Data.Tables;
using PairingApp.Models;

namespace PairingApp.Services;

public class PairingService
{
    private readonly TableClient? _tableClient;
    private readonly Dictionary<string, HashSet<string>> _pairingHistory = [];
    private readonly Dictionary<string, List<PairingEntity>> _mockStorage = [];
    private const string TableName = "pairings";

    public PairingService(TableClient? tableClient = null)
    {
        _tableClient = tableClient;
    }

    public async Task<List<Pair>> GeneratePairingsAsync(List<Participant> participants, string runDate)
    {
        Console.WriteLine($"Generating pairings for {participants.Count} participants on {runDate}");

        // Shuffle participants
        var shuffled = participants.OrderBy(_ => Guid.NewGuid()).ToList();
        var pairs = new List<Pair>();

        // Load existing pairs to avoid duplicates
        await LoadPairingHistoryAsync();

        // Generate pairs ensuring no repeats
        for (int i = 0; i < shuffled.Count - 1; i += 2)
        {
            var p1 = shuffled[i];
            var p2 = shuffled[i + 1];

            var pairKey = GetPairKey(p1.Name, p2.Name);

            // Check if this pair already exists
            if (!_pairingHistory.ContainsKey(pairKey))
            {
                pairs.Add(new Pair
                {
                    Person1 = p1.Name,
                    Person2 = p2.Name,
                    Department1 = p1.Department,
                    Department2 = p2.Department
                });

                _pairingHistory[pairKey] = new HashSet<string> { p1.Name, p2.Name };
            }
        }

        Console.WriteLine($"Created {pairs.Count} new pairs");

        // Store in Azure Table Storage
        await StorePairingsAsync(pairs, runDate);

        Console.WriteLine("Pairs stored successfully");
        return pairs;
    }

    public async Task<List<Pair>> GetPairingsAsync(string runDate)
    {
        var pairs = new List<Pair>();

        try
        {
            if (_tableClient != null)
            {
                var entities = _tableClient.QueryAsync<PairingEntity>(
                    e => e.PartitionKey == runDate
                );

                await foreach (var entity in entities)
                {
                    pairs.Add(new Pair
                    {
                        Person1 = entity.Person1 ?? string.Empty,
                        Person2 = entity.Person2 ?? string.Empty,
                        Department1 = entity.Department1 ?? string.Empty,
                        Department2 = entity.Department2 ?? string.Empty
                    });
                }
            }
            else
            {
                // Use mock storage
                if (_mockStorage.ContainsKey(runDate))
                {
                    foreach (var entity in _mockStorage[runDate])
                    {
                        pairs.Add(new Pair
                        {
                            Person1 = entity.Person1 ?? string.Empty,
                            Person2 = entity.Person2 ?? string.Empty,
                            Department1 = entity.Department1 ?? string.Empty,
                            Department2 = entity.Department2 ?? string.Empty
                        });
                    }
                }
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error retrieving pairings: {ex.Message}");
        }

        return pairs;
    }

    public async Task ResetPairingsAsync()
    {
        try
        {
            if (_tableClient != null)
            {
                var entities = _tableClient.QueryAsync<PairingEntity>();

                await foreach (var entity in entities)
                {
                    await _tableClient.DeleteEntityAsync(entity.PartitionKey, entity.RowKey);
                }
            }

            _mockStorage.Clear();
            _pairingHistory.Clear();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error resetting pairings: {ex.Message}");
            throw;
        }
    }

    public async Task<double> GetExhaustionPercentageAsync()
    {
        try
        {
            int count = 0;

            if (_tableClient != null)
            {
                var entities = _tableClient.QueryAsync<PairingEntity>();
                await foreach (var entity in entities)
                {
                    count++;
                }
            }
            else
            {
                foreach (var entities in _mockStorage.Values)
                {
                    count += entities.Count;
                }
            }

            // Simple calculation: count / (estimated max possible pairs)
            return Math.Min((double)count / 100 * 100, 100);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error calculating exhaustion: {ex.Message}");
            return 0;
        }
    }

    public double GetExhaustionPercentage()
    {
        // Sync version for quick calculation
        const int maxPossiblePairs = 100;
        var usedPairs = _pairingHistory.Count;
        return Math.Min((double)usedPairs / maxPossiblePairs * 100, 100);
    }

    public async Task<string> ExportPairingsCsvAsync(string runDate)
    {
        var pairs = await GetPairingsAsync(runDate);
        var csv = "Person 1,Department 1,Person 2,Department 2\n";

        foreach (var pair in pairs)
        {
            csv += $"\"{pair.Person1}\",\"{pair.Department1}\",\"{pair.Person2}\",\"{pair.Department2}\"\n";
        }

        return csv;
    }

    private string GetPairKey(string name1, string name2)
    {
        var names = new[] { name1, name2 }.OrderBy(n => n).ToArray();
        return $"{names[0]}|{names[1]}";
    }

    private async Task StorePairingsAsync(List<Pair> pairs, string runDate)
    {
        for (int i = 0; i < pairs.Count; i++)
        {
            var pair = pairs[i];
            var entity = new PairingEntity
            {
                PartitionKey = runDate,
                RowKey = $"pair-{i}",
                Person1 = pair.Person1,
                Person2 = pair.Person2,
                Department1 = pair.Department1,
                Department2 = pair.Department2,
                RunDate = runDate,
                Timestamp = DateTimeOffset.UtcNow
            };

            if (_tableClient != null)
            {
                try
                {
                    Console.WriteLine($"Storing pair {i} to Azure Table Storage");
                    await _tableClient.AddEntityAsync(entity);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error storing pair to Azure: {ex.Message}");
                    StoreMockEntity(entity);
                }
            }
            else
            {
                Console.WriteLine($"Using mock storage for pair {i}");
                StoreMockEntity(entity);
            }
        }
    }

    private void StoreMockEntity(PairingEntity entity)
    {
        if (!_mockStorage.ContainsKey(entity.PartitionKey))
        {
            _mockStorage[entity.PartitionKey] = [];
        }

        _mockStorage[entity.PartitionKey].Add(entity);
    }

    private async Task LoadPairingHistoryAsync()
    {
        try
        {
            if (_tableClient != null)
            {
                var entities = _tableClient.QueryAsync<PairingEntity>();
                await foreach (var entity in entities)
                {
                    var pairKey = GetPairKey(entity.Person1 ?? string.Empty, entity.Person2 ?? string.Empty);
                    _pairingHistory[pairKey] = new HashSet<string> 
                    { 
                        entity.Person1 ?? string.Empty, 
                        entity.Person2 ?? string.Empty 
                    };
                }
            }
            else
            {
                foreach (var entities in _mockStorage.Values)
                {
                    foreach (var entity in entities)
                    {
                        var pairKey = GetPairKey(entity.Person1 ?? string.Empty, entity.Person2 ?? string.Empty);
                        _pairingHistory[pairKey] = new HashSet<string> 
                        { 
                            entity.Person1 ?? string.Empty, 
                            entity.Person2 ?? string.Empty 
                        };
                    }
                }
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Error loading pairing history: {ex.Message}");
        }
    }
}

public class PairingEntity : ITableEntity
{
    public string? Person1 { get; set; }
    public string? Person2 { get; set; }
    public string? Department1 { get; set; }
    public string? Department2 { get; set; }
    public string? RunDate { get; set; }

    public string PartitionKey { get; set; } = string.Empty;
    public string RowKey { get; set; } = string.Empty;
    public DateTimeOffset? Timestamp { get; set; }
    public Azure.ETag ETag { get; set; }
}
